### PR TITLE
enable fsm in cohort

### PR DESCRIPTION
**Pull request by varks**
Enable Raft instances for the cohort fsm.

Another port for cohort raft instance. !

No changes in functionality

Subsequent PR will add:
1. concurrent map for cohort and coordinator
2. Replicate the cohort map operations similar to coordinator.

### CODE REVIEW

File: raftpb/raft.proto [hunk 1]
```diff
@@ -55,6 +60,7 @@ message RaftCommand {
 }
 
 message JoinMsg {
-    string RaftAddress  = 1;
-    string ID           = 2;
+    string RaftAddress = 1;
+    string ID = 2;
+    string TYPE = 3;
```
Comment (alex8937) [inline raftpb/raft.proto:65]: Nitpick. Maybe TYPE can be changed to a boolean for if store or cohort.

File: main.go [hunk 2]
```diff
@@ -18,29 +18,33 @@ import (
...
+	nodeID            string
+	bucketName        string
+	isCoordinator     bool
 )
 
 func init() {
 	flag.StringVarP(&listenAddress, "listen", "l", DefaultListenAddress, "Set the server listen address")
 
 	flag.StringVarP(&raftAddress, "raft", "r", DefaultRaftAddress, "Set the RAFT binding address")
+	flag.StringVarP(&cohortRaftAddress, "cohortRaft", "f", DefaultCohortRaftAddress, "Set the RAFT binding address")
```
Comment (alex8937) [inline main.go:47]: Maybe "c" is a better abbreviation. 

And default can be based on raftAddress's port + some diff for user convenience.

Comment (varks) [reply main.go:47]: c is already used for co-ordinator.

Comment (varks) [reply main.go:47]: Are you suggesting to derive cohorts raft port as a certain diff from store's raft port? Just doing it for the default won't be useful.

Comment (alex8937) [reply main.go:47]: You can derive from `&raftAddress` not DefaultRaftAddress, if unset.

Comment (varks) [reply main.go:47]: yup. I will remove the `-f` flag altogether

File: main.go [hunk 3]
```diff
@@ -79,8 +83,14 @@ func main() {
...
 		h.Start(joinHTTPAddress)
 	} else {
-		// need to start rpc server
-		kv := store.NewStore(logger, nodeID, raftAddress, raftDir, joinHTTPAddress == "", listenAddress, bucketName)
+
+		// derive raftaddress for cohort
+		port, err := strconv.ParseInt(raftAddress, 10, 32)
+		if err != nil {
+			log.Fatalf("Invalid raft port for store: %s", err)
+		}
```
Comment (alex8937) [inline main.go]: You need to split the IP address and port. Otherwise: https://play.golang.org/p/VS7YFKypFGy

Comment (varks) [reply main.go]: lol. too quick.

File: raftpb/raft.proto (hunk 1 above)
Comment (varks) [reply raftpb/raft.proto:65]: It's fine. Let's leave it this way. This is explicit.

**Comment by varks** [comment]
@alex8937 @supriya-premkumar @renga86 
If all good, I will merge it tomorrow morning. lmk

File: main.go [hunk 4]
```diff
@@ -23,6 +24,8 @@ const (
 
 	DefaultCoordinatorListenAddress = "localhost:21000"
 	DefaultCoordinatorRaftAddress   = "localhost:22000"
+
+	MagicDiff = 20000
```
Comment (supriya-premkumar) [inline main.go:28]: Can we keep this consistent? As in reserve a port if you need to. I got the coordinators come up and join the cluster inside containers. Every port exponentially makes clustering harder. We are relying on having 18 ports already :-) 
So dynamically adding an offset makes things extremely hard for containers

File: store/cohort_fsm.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+// Apply applies a Raft log entry to the key-value store.
+func (f *cohortfsm) Apply(l *raft.Log) interface{} {
+
+	var raftCommand raftpb.RaftCommand
+	if err := proto.Unmarshal(l.Data, &raftCommand); err != nil {
+		panic(fmt.Sprintf("failed to unmarshal command: %s", err.Error()))
+	}
+
+	if len(raftCommand.Commands) > 1 {
+		return fmt.Errorf("invalid command for cohort fsm")
```
Comment (supriya-premkumar) [inline store/cohort_fsm.go:26]: Format specifier missing

File: main.go (hunk 4 above)
Comment (varks) [reply main.go:28]: that is the idea. In order to not specify any ports explicitly, ports for cohorts are being derived.

Comment (alex8937) [reply main.go:28]: Let's still keep the flexibility of setting it with flag. If unset, set then port automatically then.

**Comment by alex8937** [comment]
> Enable Raft instances for the cohort fsm.
> 
> Another port for cohort raft instance. !
> 
> No changes in functionality
> 
> Subsequent PR will add:
> 
> 1. concurrent map for cohort and coordinator
> 2. Replicate the cohort map operations similar to coordinator.

I am not sure if we need to use cmap for the coordinator for the time being. The main advantage of cmap on the store fsm is it releases the global lock while waiting for the coordinator's signal of commit or abort in 2PC, whereas this doesn't exist in coordinator side. 

In the benchmark test, I saw little benefits of using cmap when assuming no network latency -- the multi-lock overhead is non-trivial in this case.

**Comment by varks** [comment]
> > Enable Raft instances for the cohort fsm.
> > Another port for cohort raft instance. !
> > No changes in functionality
> > Subsequent PR will add:
> > 
> > 1. concurrent map for cohort and coordinator
> > 2. Replicate the cohort map operations similar to coordinator.
> 
> @varks I am not sure if we need to use cmap for the coordinator for the time being. The main advantage of cmap on the store fsm is it releases the global lock while waiting for the coordinator's signal of commit or abort in 2PC, whereas this doesn't exist in coordinator side.
> 
> In the benchmark test, I saw little benefits of using cmap when assuming no network latency -- the multi-lock overhead is non-trivial in this case.

@alex8937 I agree. I made this point previously as well :)

**Pull request merged**